src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,332 @@
+   * waypoints which if the exact location is not known, an exact location is determined and chosen
+   * based on the resulting travel time. Scope of method is public for testing purposes.
+   *
+   * @param origin The starting point of travel.
+   * @param destination The ending point of travel.
+   * @param waypoints The waypoints that should be visited while travelling from the origin to the
+   *     destination.
+   * @return The most optimal set of waypoints between origin and destination.
+   * @throws GeocodingException An exception thrown when an error occurs with the Geocoding API.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   * @throws DirectionsException An exception thrown when an error occurs with the Directions API.
+   */
+  public List<String> optimizeSearchNearbyWaypoints(
+      String origin, String destination, List<String> waypoints)
+      throws GeocodingException, PlacesException, DirectionsException {
+
+    Optional<LatLng> originAsCoordinates =
+        GeocodingResultUtility.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(origin));
+
+    Optional<LatLng> destinationAsCoordinates =
+        GeocodingResultUtility.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(destination));
+
+    if (!originAsCoordinates.isPresent() || !destinationAsCoordinates.isPresent()) {
+      throw new GeocodingException("Origin or destination is invalid");
+    }
+
+    GoResponse separatedWaypoints = separateWaypoints(waypoints);
+
+    // All street address coordinates including origin and destination are collected
+    List<Optional<LatLng>> streetAddressesAsCoordinates = new ArrayList<>();
+    streetAddressesAsCoordinates.addAll(separatedWaypoints.streetAddressWaypointsAsCoordinates);
+    streetAddressesAsCoordinates.add(originAsCoordinates);
+    streetAddressesAsCoordinates.add(destinationAsCoordinates);
+
+    // Remove all optional empty entries of street address coordinates and non street address
+    // waypoints
+    List<LatLng> nonEmptyStreetAddressesAsCoordinates = filterNonNull(streetAddressesAsCoordinates);
+    List<PlaceType> nonEmptynonStreetAddressWaypointsAsPlaceTypes =
+        filterNonNull(separatedWaypoints.nonStreetAddressWaypointsAsPlaceTypes);
+
+    List<List<String>> allSearchNearbyResults =
+        searchForPlacesNearLocations(
+            nonEmptynonStreetAddressWaypointsAsPlaceTypes, nonEmptyStreetAddressesAsCoordinates);
+
+    List<List<String>> allWaypointCombinations =
+        LocationsUtility.generateCombinations(allSearchNearbyResults);
+
+    List<String> optimalWaypointCombination =
+        chooseWaypointCombinationWithShortestTravelTime(
+            origin, destination, allWaypointCombinations);
+    optimalWaypointCombination.addAll(separatedWaypoints.streetAddressWaypoints);
```

Comment:
why is this necessary?

Reply:
Oh not sure why I left that outside the function. Thanks for catching that!

Reply:
`RESOLVED`